hw/formal/tools/jaspergold/parse-fpv-report.py

```diff
@@ -0,0 +1,180 @@
+            full_file = f.read()
+            err_warn_patterns = [("errors", r"^ERROR: .*"),
+                                 ("warnings", r"^WARNING: .*"),
+                                 ("fpv_errors", r"^\[ERROR.*"),
+                                 ("fpv_warnings", r"^\[WARN.*"),
+                                 ("cex", r"^\[\d+\]\s+(\S*)\s+cex.*"),
+                                 ("undetermined", r"^\[\d+\]\s+(\S*)\s+undetermined.*"),
+                                 ("unreachable", r"^\[\d+\]\s+(\S*)\s+unreachable.*")]
+            extract_messages(full_file, err_warn_patterns, messages)
+
+            message_patterns = [("fpv_errors", r"^\[ERROR.*"),
+                                ("fpv_warnings", r"^\[WARN.*"),
+                                ("proven", r"^\[\d+\].*proven.*"),
+                                ("cex", r"^\[\d+\].*cex.*"),
+                                ("covered", r"^\[\d+\].*covered.*"),
+                                ("undetermined", r"^\[\d+\].*undetermined.*"),
+                                ("unreachable", r"^\[\d+\].*unreachable.*")]
+            extract_messages_count(full_file, message_patterns, fpv_summary)
+
+            total_msg = fpv_summary["proven"] + fpv_summary["cex"] + fpv_summary["undetermined"]
+            if total_msg > 0:
+                pass_rate = (fpv_summary["proven"] + fpv_summary["undetermined"]) / total_msg * 100
+            else:
+                pass_rate = 0
+            fpv_summary["pass_rate"] = "{0:.2f} %".format(round(pass_rate, 2))
+
+            total_msg = fpv_summary["covered"] + fpv_summary["unreachable"]
+            if total_msg > 0:
+                cov_rate = fpv_summary["covered"] / total_msg * 100
+            else:
+                cov_rate = 0
+            fpv_summary["cov_rate"] = "{0:.2f} %".format(round(cov_rate, 2))
+
+    except IOError as err:
+        messages["errors"] += ["IOError: %s" % err]
+
+    results["messages"] = messages
+    if fpv_summary:
+        results["fpv_summary"] = fpv_summary
+
+    return results
+
+
+def get_cov_results(resdir):
+    """
+    Parse FPV coverage information from the log file.
+    If no coverage information found, the function will return None
+    """
+    cov_results = OrderedDict()
+    try:
+        with Path(resdir).joinpath('fpv.log').open() as f:
+            full_file = f.read()
+            coverage_patterns = [("stimuli", r"^\|\w+.*\|(\d*.\d\d)%.*\|\d*.\d\d%.*\|\d*.\d\d%.*"),
+                                 ("coi", r"^\|\w+.*\|\d*.\d\d%.*\|(\d*.\d\d)%.*\|\d*.\d\d%.*"),
+                                 ("proof", r"^\|\w+.*\|\d*.\d\d%.*\|\d*.\d\d%.*\|(\d*.\d\d)%.*")]
+            extract_messages(full_file, coverage_patterns, cov_results)
+            for key, item in cov_results.items():
+                if len(item) == 1:
+                    cov_results[key] = item[0] + " %"
+                else:
+                    cov_results = None
+                    break
```

Comment:
This seems odd to me. Shouldn't we be spitting out a "Oh no! The log was malformed!" error message if this happens?